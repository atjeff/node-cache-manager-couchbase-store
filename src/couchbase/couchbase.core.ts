import { Store } from 'cache-manager';
import { connect } from 'couchbase';
import { InsertOptions, SetOptions, UpdateOptions } from './models/couchbase-operations.model';
import { CouchbaseStoreConfig } from './models/couchbase-store-config.model';
import { Cluster } from './types/cluster.model';

export class CouchbaseClient implements Store {
    cluster?: Cluster;
    bucket?: Bucket;
    collection!: Collection;
    config: CouchbaseStoreConfig;

    constructor(config: CouchbaseStoreConfig) {
        this.config = config;

        try {
            connect(this.config?.url, this.config).then((cluster) => {
                this.cluster = cluster;
                this.bucket = this.cluster.bucket(this.config?.bucket?.name);
                this.collection = this.bucket.defaultCollection();
            });
        } catch (e) {
            console.log('Unable to connect to couchbase. ', e);
        }
    }

    async get<T = any>(key: string): Promise<T | null | undefined> {
        if (this.isConnected()) {
            try {
                return (await this.collection.get(key)).content;
            } catch (e) {
                return null;
            }
        }
    }

    async set<T = any>(key: string, value: T, options?: SetOptions): Promise<MutationResult | void> {
        if (this.isConnected()) {
            const insertOptions: InsertOptions = this.getSetOptions(value, options);

            return await this.collection.insert(key, value, insertOptions);
        }
    }

    async del(key: string): Promise<RemoveResult | void> {
        if (this.isConnected()) {
            return await this.collection.remove(key);
        }
    }

    async upsert<T = any>(key: string, value: T, options?: UpdateOptions): Promise<MutationResult | void> {
        if (this.isConnected()) {
            const insertOptions: InsertOptions = this.getSetOptions(value, options);

            return this.collection.upsert(key, value, insertOptions);
        }
    }

    isConnected(): boolean {
        return [(this.bucket as any)?._conn?._connected, (this.collection as any)?._conn?._connected].every(Boolean);
    }

    getSetOptions<T = any>(value: T, options: SetOptions | undefined): InsertOptions {
        const insertOptions: InsertOptions = options || {};

        // We want to allow value of zero
        const ttlFactory = typeof options?.ttl !== 'undefined' ? options.ttl : this.config.ttl;

        insertOptions.timeout = typeof ttlFactory === 'function' ? ttlFactory(value) : ttlFactory;

        return insertOptions;
    }
}

export const store = {
    create: (config: CouchbaseStoreConfig) => new CouchbaseClient(config),
};
