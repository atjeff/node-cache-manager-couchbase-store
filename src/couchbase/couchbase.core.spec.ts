import { Cluster } from 'couchbase';
import { CouchbaseClient, store } from './couchbase.core';

jest.mock('couchbase');

const collectionMock = jest.fn().mockImplementation(() => ({
    _conn: {
        _connected: null,
    },
    get: jest.fn(),
    insert: jest.fn(),
    remove: jest.fn(),
    upsert: jest.fn(),
}));

const bucketMock = jest.fn().mockImplementation(() => ({
    _conn: {
        _connected: null,
    },
    defaultCollection: collectionMock,
}));

const clusterMock = jest.fn().mockImplementation(() => ({
    bucket: bucketMock,
}));

describe('CouchbaseClient', () => {
    afterEach(() => jest.clearAllMocks());

    describe('constructor', () => {
        it('should be defined', () => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);

            const couchbaseClient = new CouchbaseClient(undefined as any);

            expect(couchbaseClient).toBeDefined();
            expect(couchbaseClient).toBeInstanceOf(CouchbaseClient);
        });

        it('should set the config to the instance', () => {
            const mockConfig: any = { test: 2 };
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);

            const couchbaseClient = new CouchbaseClient(mockConfig as any);

            expect(couchbaseClient.config).toEqual(mockConfig);
        });

        it('should pass the name and config to Cluster', () => {
            const mockConfig = {
                url: 'test',
                bucket: {
                    name: 'bucket',
                    password: 'bucketPassword',
                },
            };
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);

            new CouchbaseClient(mockConfig as any);

            expect(clusterMock).toHaveBeenCalledWith(mockConfig.url, mockConfig);
        });

        it('should pass the bucket name and password to bucket', () => {
            const mockConfig = {
                url: 'test',
                bucket: {
                    name: 'bucket',
                    password: 'bucketPassword',
                },
            };
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);

            new CouchbaseClient(mockConfig as any);

            expect(bucketMock).toHaveBeenCalledWith(mockConfig.bucket.name, mockConfig.bucket.password);
        });
    });

    describe('Couchbase Methods', () => {
        let couchbaseClient: CouchbaseClient;

        beforeEach(() => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);
            couchbaseClient = new CouchbaseClient({} as any);

            jest.spyOn(couchbaseClient, 'isConnected').mockReturnValue(true);
        });

        describe('get', () => {
            it('should call collection.get', async () => {
                const mockKey = 'test';

                couchbaseClient.get(mockKey);

                expect(couchbaseClient.collection.get).toMatchSnapshot();
            });

            it('should return the value', async () => {
                const mockReturnValue = { testObj: true };
                jest.spyOn(couchbaseClient.collection, 'get').mockResolvedValue({ value: mockReturnValue });

                const result = await couchbaseClient.get('test');

                expect(result).toEqual(mockReturnValue);
            });

            it('should return null if an error is thrown', async () => {
                jest.spyOn(couchbaseClient.collection, 'get').mockRejectedValue(new Error('test error'));

                const result = await couchbaseClient.get('test');

                expect(result).toEqual(null);
            });

            it('should not call get if not connected', () => {
                jest.spyOn(couchbaseClient, 'isConnected').mockReturnValue(false);

                couchbaseClient.get('');

                expect(couchbaseClient.collection.get).not.toHaveBeenCalled();
            });
        });

        describe('set', () => {
            it('should call collection.set', async () => {
                const mockKey = 'test';
                const mockValue = { testObj: true };

                couchbaseClient.set(mockKey, mockValue);

                expect(couchbaseClient.collection.insert).toHaveBeenCalled();
            });

            it('should call getSetOptions with the passed in value and options', async () => {
                jest.spyOn(couchbaseClient, 'getSetOptions');
                const mockKey = 'test';
                const mockValue = { testObj: true };
                const mockOptions = { ttl: 5 };

                couchbaseClient.set(mockKey, mockValue, mockOptions);

                expect(couchbaseClient.getSetOptions).toMatchSnapshot();
            });

            it('should call collection.set with the result of getSetOptions', async () => {
                jest.spyOn(couchbaseClient, 'getSetOptions');
                const mockKey = 'test';
                const mockValue = { testObj: true };
                const mockOptions = { ttl: () => 5 };

                couchbaseClient.set(mockKey, mockValue, mockOptions);

                expect(couchbaseClient.collection.insert).toMatchSnapshot();
            });

            it('should not call set if not connected', () => {
                jest.spyOn(couchbaseClient, 'isConnected').mockReturnValue(false);
                const mockKey = 'test';
                const mockValue = { testObj: true };

                couchbaseClient.set(mockKey, mockValue);

                expect(couchbaseClient.collection.insert).not.toHaveBeenCalled();
            });
        });

        describe('del', () => {
            it('should call collection.del', async () => {
                const mockKey = 'test';

                couchbaseClient.del(mockKey);

                expect(couchbaseClient.collection.remove).toMatchSnapshot();
            });

            it('should not call del if not connected', () => {
                jest.spyOn(couchbaseClient, 'isConnected').mockReturnValue(false);
                const mockKey = 'test';

                couchbaseClient.del(mockKey);

                expect(couchbaseClient.collection.remove).not.toHaveBeenCalled();
            });
        });

        describe('upsert', () => {
            it('should call collection.upsert', async () => {
                const mockKey = 'test';
                const mockValue = { testObj: true };

                couchbaseClient.upsert(mockKey, mockValue);

                expect(couchbaseClient.collection.upsert).toMatchSnapshot();
            });

            it('should call getSetOptions with the passed in value and options', async () => {
                jest.spyOn(couchbaseClient, 'getSetOptions');
                const mockKey = 'test';
                const mockValue = { testObj: true };
                const mockOptions = { ttl: 5 };

                couchbaseClient.upsert(mockKey, mockValue, mockOptions);

                expect(couchbaseClient.getSetOptions).toMatchSnapshot();
            });

            it('should call collection.upsert with the result of getSetOptions', async () => {
                jest.spyOn(couchbaseClient, 'getSetOptions');
                const mockKey = 'test';
                const mockValue = { testObj: true };
                const mockOptions = { ttl: () => 5 };

                couchbaseClient.upsert(mockKey, mockValue, mockOptions);

                expect(couchbaseClient.collection.upsert).toMatchSnapshot();
            });

            it('should not call upsert if not connected', () => {
                jest.spyOn(couchbaseClient, 'isConnected').mockReturnValue(false);
                const mockKey = 'test';
                const mockValue = { testObj: true };

                couchbaseClient.upsert(mockKey, mockValue);

                expect(couchbaseClient.collection.upsert).not.toHaveBeenCalled();
            });
        });
    });

    describe('isConnected', () => {
        it('should return the result of Array.some', () => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);
            const couchbaseClient: CouchbaseClient = new CouchbaseClient({} as any);

            [
                {
                    bucket: true,
                    collection: true,
                    result: true,
                },
                {
                    bucket: true,
                    collection: false,
                    result: false,
                },
                {
                    bucket: false,
                    collection: true,
                    result: false,
                },
                {
                    bucket: false,
                    collection: false,
                    result: false,
                },
                {
                    bucket: undefined,
                    collection: undefined,
                    result: false,
                },
            ].forEach(({ bucket, collection, result }) => {
                (couchbaseClient.bucket as any)._conn._connected = bucket;
                couchbaseClient.collection._conn._connected = collection;

                expect(couchbaseClient.isConnected()).toEqual(result);
            });
        });

        it('should default values to undefined if not connected', () => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);
            const couchbaseClient: CouchbaseClient = new CouchbaseClient({} as any);

            delete (couchbaseClient.bucket as any)._conn;
            delete couchbaseClient.collection._conn;

            expect(couchbaseClient.isConnected()).toEqual(false);
        });

        it('should default values if defaultCollection or bucket fails', () => {
            collectionMock.mockReturnValue(undefined);
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);
            const couchbaseClient: CouchbaseClient = new CouchbaseClient({} as any);

            expect(couchbaseClient.isConnected()).toEqual(false);
        });
    });

    describe('getSetOptions', () => {
        let couchbaseClient: CouchbaseClient;

        beforeEach(() => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);
            couchbaseClient = new CouchbaseClient({} as any);
        });

        it('should call the ttl function if it exists', () => {
            const mockTtl = jest.fn().mockReturnValue(5);

            couchbaseClient.getSetOptions(null, { ttl: mockTtl });

            expect(mockTtl).toHaveBeenCalled();
        });

        it('should set the ttl function result if it exists', () => {
            const mockTtl = jest.fn().mockReturnValue(5);

            const result = couchbaseClient.getSetOptions(null, { ttl: mockTtl });

            expect(result).toMatchSnapshot();
        });

        it('should set the ttl value if it exists', () => {
            const mockTtl = 12;

            const result = couchbaseClient.getSetOptions(null, { ttl: mockTtl });

            expect(result).toMatchSnapshot();
        });
    });

    describe('create', () => {
        it('should create a new CouchbaseClient with the passed in config', () => {
            (Cluster as jest.MockedClass<typeof Cluster>).mockImplementation(clusterMock);

            const couchbaseClient = store.create({} as any);

            expect(couchbaseClient).toBeDefined();
            expect(couchbaseClient).toBeInstanceOf(CouchbaseClient);
        });
    });
});
