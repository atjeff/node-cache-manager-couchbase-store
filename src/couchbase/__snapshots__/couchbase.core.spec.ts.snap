// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CouchbaseClient Couchbase Methods del should call collection.del 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods get should call collection.get 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods set should call collection.set with the result of getSetOptions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test",
      Object {
        "testObj": true,
      },
      Object {
        "timeout": 5,
        "ttl": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods set should call getSetOptions with the passed in value and options 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "testObj": true,
      },
      Object {
        "timeout": 5,
        "ttl": 5,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "timeout": 5,
        "ttl": 5,
      },
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods upsert should call collection.upsert 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test",
      Object {
        "testObj": true,
      },
      Object {
        "timeout": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods upsert should call collection.upsert with the result of getSetOptions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test",
      Object {
        "testObj": true,
      },
      Object {
        "timeout": 5,
        "ttl": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CouchbaseClient Couchbase Methods upsert should call getSetOptions with the passed in value and options 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "testObj": true,
      },
      Object {
        "timeout": 5,
        "ttl": 5,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "timeout": 5,
        "ttl": 5,
      },
    },
  ],
}
`;

exports[`CouchbaseClient getSetOptions should set the ttl function result if it exists 1`] = `
Object {
  "timeout": 5,
  "ttl": [MockFunction] {
    "calls": Array [
      Array [
        null,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": 5,
      },
    ],
  },
}
`;

exports[`CouchbaseClient getSetOptions should set the ttl value if it exists 1`] = `
Object {
  "timeout": 12,
  "ttl": 12,
}
`;
